substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+	///
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* this method is
+	/// called.
+	fn on_nominator_add(who: &T::AccountId, nominations: Vec<AccountIdOf<T>>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		// voter may exist in the list in case of re-enabling a chilled nominator;
+		if T::VoterList::contains(who) {
+			return
+		}
+
+		let _ = T::VoterList::on_insert(who.clone(), nominator_vote)
+			.defensive_proof("staker does not exist in the list as per check above; qed.");
+
+		// If who is a nominator, update the vote weight of the nominations if they exist. Note:
+		// this will update the score of up to `T::MaxNominations` validators.
+		match T::Staking::status(who).defensive() {
+			Ok(StakerStatus::Nominator(_)) =>
+				for t in nominations {
+					Self::update_target_score(&t, StakeImbalance::Positive(nominator_vote.into()))
+				},
+			Ok(StakerStatus::Idle) | Ok(StakerStatus::Validator) | Err(_) => (), // nada.
+		};
+
+		log!(debug, "on_nominator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+	}
+
+	/// Fired when a nominator becomes idle (i.e. chilling).
+	///
+	/// From the `T::VotertList` PoV, chilling a nominator is the same as removing it.
+	///
+	/// Note: it is assumed that `who`'s staking ledger and `nominations` are up to date before
+	/// calling this method.
+	fn on_nominator_idle(who: &T::AccountId, nominations: Vec<T::AccountId>) {
+		Self::on_nominator_remove(who, nominations);
+	}
+
+	/// Fired when someone removes their intention to nominate and is completely removed from
+	/// the staking state.
+	///
+	/// Note: the number of nodes that are updated is bounded by the maximum number of
+	/// nominators, which is defined in the staking pallet.
+	fn on_nominator_remove(who: &T::AccountId, nominations: Vec<T::AccountId>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		log!(
+			debug,
+			"remove nominations from {:?} with {:?} weight. impacting {:?}.",
+			who,
+			nominator_vote,
+			nominations,
+		);
+
+		// updates the nominated target's score.
+		for t in nominations.iter() {
+			Self::update_target_score(t, StakeImbalance::Negative(nominator_vote.into()))
+		}
+
+		let _ = T::VoterList::on_remove(who)
+			.defensive_proof("the nominator exists in the list as per the contract with staking.");
+	}
+
+	/// Fired when an existing nominator updates their nominations.
+	///
+	/// This is called when a nominator updates their nominations. The nominator's stake remains
+	/// the same (updates to the nominator's stake should emit [`Self::on_stake_update`]
+	/// instead). However, the score of the nominated targets must be updated accordingly.
+	///
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* calling this
+	/// method.
+	fn on_nominator_update(
+		who: &T::AccountId,
+		prev_nominations: Vec<T::AccountId>,
+		nominations: Vec<AccountIdOf<T>>,
```

Comment:
Isn't this just a `Vec<T::AccountId>` too? Why is it different?

Reply:
Yes, it's the same type. I'll change to `T::Account` for readability.